libraries/AP_Torqeedo/AP_Torqeedo.h

```diff
@@ -0,0 +1,111 @@
+ */
+ 
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+
+#ifndef TORQEEDO_ENABLED
+#define TORQEEDO_ENABLED !HAL_MINIMIZE_FEATURES
+#endif
+
+#if TORQEEDO_ENABLED
+
+#include <AP_Param/AP_Param.h>
+
+#define TORQEEDO_MESSAGE_LEN_MAX    30  // messages are no more than 30 bytes
+
+class AP_Torqeedo {
+public:
+    AP_Torqeedo();
+
+    /* Do not allow copies */
+    AP_Torqeedo(const AP_Torqeedo &other) = delete;
+    AP_Torqeedo &operator=(const AP_Torqeedo&) = delete;
+
+    static AP_Torqeedo* get_singleton();
+
+    // initialise driver
+    void init();
+
+    // consume incoming messages from motor, reply with latest motor speed
+    // runs in background thread
+    void loop();
+
+    static const struct AP_Param::GroupInfo var_info[];
+
+private:
+
+    // message ids
+    enum class MsgId : uint8_t {
+        SET_MOTOR_SPEED = 0x0,
+        UNKNOWN_0x01 = 0x1,
+        REQUEST_MOTOR_SPEED = 0x14,
+        UNKNOWN_0x20 = 0x20,
+        UNKNOWN_0x30 = 0x30,
+    };
+
+    enum class ParseState {
+        WAITING_FOR_HEADER = 0,
+        WAITING_FOR_FOOTER,
+    };
+
+    // initialise serial port and gpio pins (run from background thread)
+    void init_internals();
+
+    // process a single byte received on serial port
+    // return true if a this driver should send a set-motor-speed message
+    bool parse_byte(uint8_t b);
+
+    // set pin to enable sending commands to motor
+    void send_start();
+
+    // check for timeout after sending and unset pin if required
+    void check_for_send_end();
+
+    // calculate delay require to allow bytes to be sent
+    uint32_t calc_send_delay_us(uint8_t num_bytes);
+
+    // send a motor speed command as a value from -1000 to +1000
+    // value is taken directly from SRV_Channel
+    void send_motor_speed_cmd();
+
+    // parameters
+    AP_Int8 _enable;        // 1 if torqeedo feature is enabled
+    AP_Int8 _pin_onoff;     // Pin number connected to Torqeedo's on/off pin. -1 to disable turning motor on/off from autopilot
+    AP_Int8 _pin_de;        // Pin number connected to RS485 to Serial converter's DE pin. -1 to disable sending commands to motor
+
+    // members
+    AP_HAL::UARTDriver *_uart;      // serial port to communicate with motor
+    bool _initialised;              // true once driver has been initialised
+    int16_t _motor_speed;           // desired motor speed (set from within update method)
+    uint32_t _last_send_motor_us;   // system time (in micros) last motor speed command was sent
+    uint32_t _send_delay_us;        // delay (in micros) to allow bytes to be sent after which pin can be unset.  0 if not delaying
+
+    // message parsing members
+    ParseState _parse_state;        // current state of parsing
+    uint32_t parse_bytes_received;  // number of bytes received
+    uint32_t parse_error_count;     // total number of parsing errors (for reporting)
+    uint32_t parse_success_count;   // number of messages successfully parsed (for reporting)
```

Comment:
None of these are used, but a simple `Log_Write("TQDO", ...` might be useful.

Reply:
Yes, I'm planning to add logging in the near-ish future.  I know we don't like to add stuff that isn't being used but it will be soon and I won't forget.. it definitely really will be used soon.